File: MD/C/control.c
/*
 * $Id: control-c.c,v 1.2 2002/01/08 12:32:48 spb Exp spb $
 *
 * Control program for the MD update
 *
 */
#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h>

#define DECL
#include "coord.h"

double second(void); 
int main(int argc, char *argv[]){
  int i,j;
FILE *in, *out;
double start,stop;
 char name[80];
/*  timestep value */
double dt=0.02;

/*  number of timesteps to use. */
int Nstep=100;
 int Nsave=5;
  if( argc > 1 ){
   Nstep=atoi(argv[1]);
  }

  /* set up multi dimensional arrays */
  r = calloc(Nbody,sizeof(double));
  delta_r = calloc(Nbody*Nbody,sizeof(double));
  mass = calloc(Nbody,sizeof(double));
  visc = calloc(Nbody,sizeof(double));
  f[0] = calloc(Ndim*Nbody,sizeof(double));
  pos[0] = calloc(Ndim*Nbody,sizeof(double));
  vel[0] = calloc(Ndim*Nbody,sizeof(double));
  delta_x[0] = calloc(Ndim*Nbody*Nbody,sizeof(double));
  for(i=1;i<Ndim;i++){
    f[i] = f[0] + i * Nbody;
    pos[i] = pos[0] + i * Nbody;
    vel[i] = vel[0] + i * Nbody;
    delta_x[i] = delta_x[0] + i*Nbody*Nbody;
  }

/* read the initial data from a file */

  collisions=0;
  in = fopen("input.dat","r");

  if( ! in ){
    perror("input.dat");
    exit(1);
  }

  for(i=0;i<Nbody;i++){
    fscanf(in,"%16le%16le%16le%16le%16le%16le%16le%16le\n",mass+i,visc+i,
      &pos[Xcoord][i], &pos[Ycoord][i], &pos[Zcoord][i],
      &vel[Xcoord][i], &vel[Ycoord][i], &vel[Zcoord][i]);
  }
  fclose(in);

/*
 * Run Nstep timesteps and time how long it took
 */

  for(j=1;j<=Nsave;j++){
      start=second();
      evolve(Nstep,dt); 
      stop=second();


      printf("%d timesteps took %f seconds\n",j*Nstep,stop-start);
      printf("collisions %d\n",collisions);

/* write final result to a file */
      sprintf(name,"output.dat%03d",j*Nstep);
      out = fopen(name,"w");

      if( ! out ){
	perror(name);
	exit(1);
      }

      for(i=0;i<Nbody;i++){
	fprintf(out,"%16.8E%16.8E%16.8E%16.8E%16.8E%16.8E%16.8E%16.8E\n",
		mass[i],visc[i],
		pos[Xcoord][i], pos[Ycoord][i], pos[Zcoord][i],
		vel[Xcoord][i], vel[Ycoord][i], vel[Zcoord][i]);
      }
      fclose(out);
  }

}

double second()
{
/* struct timeval { long        tv_sec; 
            long        tv_usec;        };

struct timezone { int   tz_minuteswest;
             int        tz_dsttime;      };     */

        struct timeval tp;
        struct timezone tzp;
        int i;

        i = gettimeofday(&tp,&tzp);
        return ( (double) tp.tv_sec + (double) tp.tv_usec * 1.e-6 );
}

File: MD/Fortran/control.f
C
C $Id: control.f,v 1.1 2002/01/08 12:30:07 spb Exp spb $
C
C Control program for the MD update
C

      PROGRAM MD
      IMPLICIT NONE
      INCLUDE 'coord.inc'
      INTEGER i,j
      REAL start, stop
      REAL tarray(2)
      REAL etime

C  timestep value
      DOUBLE PRECISION dt
      PARAMETER( dt = 0.02 )

C  number of timesteps to use.
      INTEGER Nstep
      INTEGER Nsave
      PARAMETER(Nstep=100,Nsave=5)
      CHARACTER*80 name

C read the initial data from a file

      collisions=0
      OPEN(unit=1,file='input.dat',status='OLD')
      DO i=1,Nbody
        READ(1,10) mass(i),visc(i),
     $    pos(i,Xcoord),pos(i,Ycoord),pos(i,Zcoord),
     $    vel(i,Xcoord),vel(i,Ycoord),vel(i,Zcoord)
      END DO
      CLOSE(unit=1)

C
C Run 20 timesteps and time how long it took
C


      start = etime(tarray)
      DO j=1,Nsave
        CALL evolve(Nstep,dt)
        stop = etime(tarray)

        WRITE(*,*) (j*Nstep), ' timesteps took ',(stop-start)
        write(*,*) collisions, ' collisions'


C write result to a file
        WRITE(name,12) 'output.dat', j*Nstep
        OPEN(unit=1,file=name)
        DO i=1,Nbody
          WRITE(1,10) mass(i),visc(i),
     $      pos(i,Xcoord),pos(i,Ycoord),pos(i,Zcoord),
     $      vel(i,Xcoord),vel(i,Ycoord),vel(i,Zcoord)
        END DO
        CLOSE(unit=1)
      END DO

10    FORMAT(8E16.8)
12    FORMAT(A10,I3.3)

      END

